gawkextlib: init at unstable

tomberek opened the pull request:
###### Motivation for this change
Using AWK recently. Found some extensions that are useful for JSON, etc.

###### Things done


- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

tomberek committed:
gawkextlib: init at unstable

Update with suggestions from veprbl

Update

Refactor gawk extensions

Use setupHook mechanism to produce AWKLIBPATH, though still uses a
wrapper in the end. Independently builds extensions. Provide example
gawk-with-extensions to show use.

veprbl reviewed (commented):
Looks good. Some parts could use some rework.

tomberek commented on pkgs/tools/text/gawk/gawk-with-extensions.nix:
```diff
@@ -0,0 +1,12 @@
+buildEnv rec {
+    name = "${gawk.name}-with-extensions";
+    paths = [ gawk ] ++ extensions;
+    pathsToLink = [ "/bin" "/etc" "/include" "/lib" "/libexec" "/share" ];
```
This pulls in all of gawk to a single directory. There is probably a cleaner way to accomplish this, but it's what I was able to get working.

tomberek commented on pkgs/tools/text/gawk/gawk-with-extensions.nix:
```diff
@@ -0,0 +1,12 @@
+    pathsToLink = [ "/bin" "/etc" "/include" "/lib" "/libexec" "/share" ];
+    buildInputs = [ makeWrapper ];
+    postBuild = ''
+      cp $out/lib/gawk/* $out/lib/.
```
Yep, good call... fixing..

tomberek commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,80 @@
+
+  configurePhase = "";
+  postPatch = ''
+    cd lib
```
We need the normal autoreconf to run in the lib directory, buildPhase is too late.

tomberek commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,80 @@
+  postPatch = ''
+    cd lib
+  '';
+  buildPhase = ''
```
That was my first try, but ran into some impurities. If someone could figure out the right way to patch that process into success, that would be great.

tomberek commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,80 @@
+        are useful for building gawk extension libraries. These libraries
+        enable gawk to process XML data, interact with a PostgreSQL
+        database, use the GD graphics library, and perform unlimited
+        precision MPFR calculations.
```
That might be an eventual outcome. I don't have a good idea yet of whether or not people would even use gawkextlib, so for now I just left it bundled. The main advantage I see at the moment is that Nix provides a much easier mechanism for people to try these extensions out.

tomberek commented:
@veprbl Thanks for the review. Updated PR taking your suggestions. There are a few I don't have a solid answer for at the moment.

mmahut commented:
Are there any updates on this pull request, please?

tomberek commented:
Not right now. I’m willing to make some changes if there is interest.

tomberek commented:
Woops. Did a massive merge by mistake. Force-pushed a more reasonable change set.

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3393,6 +3393,12 @@ in
   };
 
+  gawk-with-extensions = callPackage ../tools/text/gawk/gawk-with-extensions.nix {
+    extensions = [ gawkextlib ];
```
This doesn't work anymore since gawkextlib is now a wrapper for `gawk`.

veprbl reviewed (commented):
I think this PR needs at least following changes:
* [ ] get rid of the buildEnv because it is not composable
* [ ] split different extensions into separate derivations with different dependencies (maybe define some `buildGawkextlibExtension` to factor out the common parts)
* [ ] switch from using wrappers to a `setupHook` for each extension that would append "$out/lib" to the AWKLIBPATH. This is not a must, but, for packaging plugins, the hooks are generally preferred in nixpkgs over wrappers.

tomberek commented:
@veprbl Attempted a refactor. Not quite sure how to use setupHook and still made a wrapper. Though it does use the setupHook's to build the AWKPATHLIB. Let me know if this is on the right track.

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+let
+  buildGawkextlibExtension = extension: extraDeps:
+    stdenv.mkDerivation rec {
+      name = "gawkextlib-${extension}-2019-11-21";
```
```suggestion
      pname = "gawkextlib-${extension}";
      version = "unstable-2019-11-21";
```

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+      src = fetchgit {
+        url = "git://git.code.sf.net/p/gawkextlib/code";
+        rev = "f70f10da2804e4fd0a0bac57736e9c1cf21e345d";
+        sha256 = "sha256-M3bBjOp8OrrOosEDScEgJUEFJPYApaC/do3QYRP6DmU=";
```
```suggestion
        sha256 = "0r8fz89n3l4dfszs1980yqj0ah95430lj0y1lb7blfkwxa6c2xik";
```

veprbl commented on pkgs/tools/text/gawk/gawk-with-extensions.nix:
```diff
@@ -0,0 +1,12 @@
+{ runCommand, gawk, extensions ? [ ], makeWrapper }:
```
```suggestion
{ runCommand, gawk, extensions ? [], makeWrapper }:
```

veprbl replied:
Or even remove the default value.

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+      cp ../lib/gawkextlib.h $out/lib/.
+    '';
+  });
+in {
```
```suggestion
in recurseIntoAttrs {
```

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+in {
+
+  # callPackage injects extra items into the root attrSet
+  full = {
```
`gawkextlib-extensions.full.json` would look a bit strange. Maybe just reduce nesting and define `full` in some pedestrian way:
```
full = [ abort aregex csv errno gd haru json lmdb mbs mpfr nl_langinfo pgsql select xml timex ];
```

tomberek replied:
Returning this as a simple list seems to cause an infinite recursion.

tomberek replied:
Resolved by filtering at the point of use.

veprbl commented on pkgs/tools/text/gawk/gawk-with-extensions.nix:
```diff
@@ -0,0 +1,12 @@
+  for i in ${gawk}/bin/*; do
+    name="$(basename "$i")"
+    makeWrapper $i $out/bin/$name \
+      --suffix AWKLIBPATH : "${gawk}/lib/gawk:''${AWKLIBPATH:-}"
```
`--suffix` means that the wrapper will always override user setting for AWKLIBPATH?

tomberek replied:
```suggestion
      --prefix AWKLIBPATH : "${gawk}/lib/gawk:''${AWKLIBPATH:-}"
```

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+      preCheck = ''
+        AWKLIBPATH=${gawk}/lib/gawk
+      '';
+      setupHook = writeText "setupHook.sh" ''
```
Can we have a real "setup-hook.sh" file instead?

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+      postPatch = ''
+        cd ${extension}
+      '';
+      installPhase = ''
```
Missing `runHook preInstall`

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+        gawk
+        gawkextlib
+        #rapidjson gd shapelib libharu
+        #lmdb gmp mpfr postgresql hiredis expat tre
```
Needs cleanup

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3393,6 +3393,12 @@ in
   };
 
+  gawk-with-extensions = callPackage ../tools/text/gawk/gawk-with-extensions.nix {
+    inherit gawk;
```
This line is not needed.

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,101 @@
+      };
+    };
+  gawkextlib = (buildGawkextlibExtension "lib" []).overrideAttrs (old: {
+    configureFlags = null;
```
```suggestion
    configureFlags = [];
```
Setting an attribute to `null` in a derivation does not undefine it.

veprbl reviewed (commented):
This looks great

tomberek committed:
Apply suggestions from code review

Co-Authored-By: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>

Commit:
Clean up extension implementation.

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3393,6 +3393,13 @@ in
 
+  gawk-with-extensions = callPackage ../tools/text/gawk/gawk-with-extensions.nix {
+    extensions = let
+      gawkextlib = callPackage ../tools/text/gawk/gawkextlib.nix {};
```
This is a big step back. This prevents users from using extensions with setupHook like
```
nix-shell -p gawk -p gawk-extensions.json
```
And it also prevents users from doing `gawk-with-extensions.override { extensions = [ gawkextlib.json ] }` because you hid the orignal value of `gawkextlib`.

tomberek replied:
ah, okay, reverting. I was trying to avoid polluting top-level

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,97 @@
+
+      # Needed for some tests
+      preCheck = ''
+        AWKLIBPATH=${gawk}/lib/gawk
```
Shouldn't setupHook of `gawkextlib` handle this?

tomberek replied:
No. gawk itself should. This causes a problem when trying to do `nix-shell --pure -p gawk -p gawkextlib.json` i think gawk does some relative path searching when AWKLIBPATH is not defined, but as soon as you start defining it (and the tests in the select extension do), you need that in the path. 

Not sure if I should just add a setupHook to gawk itself.

tomberek replied:
Added the `${gawk}/lib/gawk` piece to a the gawkextlib setupHook only. This means it is inherited by the actual extensions and enables the use of
```
nix-shell -I nixpkgs=. --pure -p gawk -p gawkextlib.json
```
without breaking things like time.so and filefuncs.so (these come with gawk itself).

This made a lot of other things much simpler at the cost of two separate setupHooks.

veprbl replied:
Maybe you can use gawklibpath_append? (should be available after you source @gawk@/etc/profile.d/gawk.sh, need to set `gawk` attribute for @gawk@ to work)
It looks like it will not work with set -u, though. You might need to do
```
    local oldOpts="-u"
    shopt -qo nounset || oldOpts="+u"
    set +u

    set "$oldOpts"
```
or just copy paste their implementation into our setupHook.

veprbl commented on pkgs/tools/text/gawk/gawkextlib.nix:
```diff
@@ -0,0 +1,97 @@
+in recurseIntoAttrs {
+
+  # callPackage injects extra items into the root attrSet
+  extensions = {
```
I still don't get the purpose of this extra nesting.

veprbl replied:
> callPackage injects extra items into the root attrSet

True, is that an issue?

tomberek replied:
I think callPackage puts "override" and "overrideDerivation" into the set. Otherwise i can just import, but then we don't get the callPackage magic. Then if it's an attrSet, it needs to turn into a list somewhere, either in top-level, or once it goes to gawk-with-extensions.nix, either way, the extra items need to be removed.

New version just filters those out at the point of use.

Commit:
Cleanup nested structure. Limit overrides only to when needed

Commit:
Core gawkextlib also sets up gawk build-in lib

Commit:
Simplify setuphook. Re-use existing gawk.sh

tomberek commented:
Anything else needed?

veprbl committed:
add recurseIntoAttrs

veprbl committed:
s/extraDeps/extraBuildInputs/

veprbl committed:
add libiconv to xml

Fixes build error (on darwin?)

In file included from xml_puller.c:18:
./xml_puller.h:17:10: fatal error: 'iconv.h' file not found
         ^~~~~~~~~
1 error generated.

veprbl committed:
implement gawkextlib.full

veprbl committed:
move more to checkInputs

veprbl committed:
s/buildGawkextlibExtension/buildExtension/

veprbl committed:
make gawkextlib overridable, get rid of overrideAttrs in the definition

veprbl committed:
enable tests on darwin

veprbl committed:
run nixfmt

veprbl committed:
remove custom installPhase, make install should work just fine

veprbl committed:
remove configureFlags

veprbl committed:
(refactor) move postPatch around

veprbl committed:
gawk-with-extensions.nix: lib is not used

veprbl commented:
@tomberek I made some changes, could you review?

tomberek commented:
Builds and functions as expected

veprbl committed:
use alphabetic order

veprbl commented:
@GrahamcOfBorg build gawkextlib gawk-with-extensions

veprbl committed:
make buildExtension public

veprbl commented:
@GrahamcOfBorg build gawkextlib gawk-with-extensions

veprbl reviewed (approved):
`gawkextlib.json` appears to be working
